program/src/state/v2/rule_set_v2.rs

```diff
@@ -0,0 +1,254 @@
+    pub fn size(&self) -> u32 {
+        self.header[1]
+    }
+
+    /// Deserialize a constraint from a byte array.
+    pub fn from_bytes(bytes: &'a [u8]) -> Result<Self, RuleSetError> {
+        // header
+        let header = try_from_bytes::<[u32; 2]>(0, U64_BYTES, bytes)?;
+        let mut cursor = U64_BYTES;
+
+        // owner
+        let owner = try_from_bytes::<Pubkey>(cursor, PUBKEY_BYTES, bytes)?;
+        cursor += PUBKEY_BYTES;
+
+        // name
+        let rule_set_name = try_from_bytes::<Str32>(cursor, Str32::SIZE, bytes)?;
+        cursor += Str32::SIZE;
+
+        // number of operations and rules
+        let size = header[1] as usize;
+
+        // operations
+        let slice_end = cursor
+            + Str32::SIZE
+                .checked_mul(size)
+                .ok_or(RuleSetError::NumericalOverflow)?;
+
+        // sanity check: make sure we got the correct slice size
+        if (slice_end + 1) > bytes.len() {
+            msg!("Invalid slice end: {} > {}", slice_end, bytes.len());
+            return Err(RuleSetError::DeserializationError);
+        }
+
+        let operations = try_cast_slice(&bytes[cursor..slice_end])?;
+        cursor = slice_end;
+
+        // rules
+        let mut rules = Vec::with_capacity(size);
+
+        for _ in 0..size {
+            let rule = RuleV2::from_bytes(&bytes[cursor..]).unwrap();
+            cursor += rule.length();
+            rules.push(rule);
+        }
+
+        Ok(Self {
+            header,
+            owner,
+            rule_set_name,
+            operations,
+            rules,
+        })
+    }
+
+    /// Serialize a `RuleSetV2` into a byte array.
+    pub fn serialize(
+        owner: Pubkey,
+        name: &str,
+        operations: &[String],
+        rules: &[&[u8]],
+    ) -> Result<Vec<u8>, RuleSetError> {
+        // length of the rule set
+        let length = U64_BYTES
+            + PUBKEY_BYTES
+            + Str32::SIZE
+            + (operations.len() * Str32::SIZE)
+            + rules
+                .iter()
+                .map(|v| v.len())
+                .reduce(|accum, item| accum + item)
+                .ok_or(RuleSetError::DataIsEmpty)
+                .unwrap();
+
+        let mut data = Vec::with_capacity(length);
+
+        // header section
+        // - lib version
+        data.extend([LibVersion::V2 as u8, 0, 0, 0]);
+        // - size
+        data.extend(u32::to_le_bytes(operations.len() as u32));
+
+        // owner
+        data.extend(owner.as_ref());
+
+        // name
+        let mut field_bytes = [0u8; Str32::SIZE];
+        field_bytes[..name.len()].copy_from_slice(name.as_bytes());
+        data.extend(field_bytes);
+
+        // operations
+
+        // sanity check: checks whether we have duplicated operation names
+        if (1..operations.len()).any(|i| operations[i..].contains(&operations[i - 1])) {
+            return Err(RuleSetError::DuplicatedOperationName);
+        }
```

Comment:
Yeah, this is not the most efficient way to do it but (1) this code does not run on-chain, only on the client; (2) only once during the rule set creation; and (3) the number of operations is relatively small.

On-chain there is similar procedure to find the operation:
```
let mut bytes = [0u8; Str32::SIZE];
bytes[..operation.len()].copy_from_slice(operation.as_bytes());

for (i, operation) in self.operations.iter().enumerate() {
        if sol_memcmp(&operation.value, &bytes, bytes.len()) == 0 {
                return Some(&self.rules[i]);
        }
}
```
I still think this is more efficient than using a `HashSet`, since creating the `HashSet` on-chain will be expensive and we are doing a `memcpm`, which likely is more efficient than `O(n)`. So in the end, this will be somewhere between `O(n)` and `O(n^2)`, but should be more efficient than `O(n^2)`.